File: app/src/main/java/com/metahumanz/pacilread/HomeNavigationController.java
package com.metahumanz.pacilread;

import android.app.Activity;
import android.view.MotionEvent;
import android.view.View;

import com.metahumanz.pacilread.storage.SettingsStore;

import java.util.ArrayList;
import java.util.List;

public final class HomeNavigationController {
    public static final int PAGE_BOOKSHELF = 0;
    public static final int PAGE_STATS = 1;
    public static final int PAGE_BOOKMARKS = 2;
    public static final int PAGE_SETTINGS = 3;

    public interface Callback {
        boolean isReadingTimeTrackingEnabled();
        void onHomePageSelected(int page, boolean syncFirst);
    }

    private final Activity activity;
    private final Callback callback;
    private final List<Integer> activePages = new ArrayList<>();
    private final View rootView;
    private final View pageContainer;
    private final View sectionBookshelf;
    private final View sectionReadingStats;
    private final View sectionBookmarks;
    private final View sectionSettings;
    private final View loadingLayout;
    private final View sidebarBookshelfButton;
    private final View sidebarStatsButton;
    private final View sidebarBookmarksButton;
    private final View sidebarSettingsButton;
    private final HomeNavigationResolver resolver;
    private final HomeBottomNavigationController bottomNavigationController;
    private final HomeSidebarController sidebarController;
    private final HomePagerController pagerController;

    private int currentPage = PAGE_BOOKSHELF;
    private int effectiveMode = HomeNavigationResolver.MODE_BOTTOM;

    public HomeNavigationController(Activity activity, SettingsStore settingsStore, Callback callback) {
        this.activity = activity;
        this.callback = callback;
        this.rootView = activity.findViewById(R.id.main_root);
        this.pageContainer = activity.findViewById(R.id.page_container);
        this.sectionBookshelf = activity.findViewById(R.id.section_bookshelf);
        this.sectionReadingStats = activity.findViewById(R.id.section_reading_stats);
        this.sectionBookmarks = activity.findViewById(R.id.section_bookmarks);
        this.sectionSettings = activity.findViewById(R.id.section_home_settings);
        this.loadingLayout = activity.findViewById(R.id.layout_loading);
        this.sidebarBookshelfButton = activity.findViewById(R.id.button_home_sidebar_bookshelf);
        this.sidebarStatsButton = activity.findViewById(R.id.button_home_sidebar_stats);
        this.sidebarBookmarksButton = activity.findViewById(R.id.button_home_sidebar_bookmarks);
        this.sidebarSettingsButton = activity.findViewById(R.id.button_home_sidebar_settings);
        this.resolver = new HomeNavigationResolver(activity, settingsStore, rootView);
        this.bottomNavigationController = new HomeBottomNavigationController(
                activity,
                settingsStore,
                resolver,
                page -> selectHomePage(page, true, page == PAGE_STATS)
        );
        this.sidebarController = new HomeSidebarController(
                activity,
                settingsStore,
                resolver,
                (page, animate) -> selectHomePage(page, animate, page == PAGE_STATS)
        );
        this.pagerController = new HomePagerController(pageContainer, new HomePagerController.Callback() {
            @Override
            public View sectionForPage(int page) {
                return HomeNavigationController.this.sectionForPage(page);
            }

            @Override
            public boolean isTouchBlocked(MotionEvent event) {
                return loadingLayout == null
                        || loadingLayout.getVisibility() == View.VISIBLE
                        || bottomNavigationController.isTouchInBottomNavigation(event);
            }

            @Override
            public void onPageChanged(int page, boolean syncFirst) {
                currentPage = page;
                callback.onHomePageSelected(page, syncFirst || page == PAGE_STATS);
            }

            @Override
            public void onSelectionChanged() {
                updateNavigationSelection();
            }
        });
        bindSidebarMenuButtons();
        setup();
    }

    public int getCurrentPage() {
        return currentPage;
    }

    public void restoreHomePage(int page) {
        selectHomePage(page, false, false);
    }

    public void refreshFromSettings() {
        activePages.clear();
        activePages.addAll(resolver.activePages(callback.isReadingTimeTrackingEnabled()));
        int previousMode = effectiveMode;
        effectiveMode = resolver.resolveEffectiveMode();
        if (!activePages.contains(currentPage)) {
            currentPage = PAGE_BOOKSHELF;
        }
        pagerController.setActivePages(activePages);
        pagerController.setCurrentPage(currentPage);
        sidebarController.setActivePages(activePages);
        bottomNavigationController.update(activePages, currentPage, effectiveMode);
        sidebarController.updateShell(effectiveMode, previousMode != effectiveMode);
        sidebarController.rebuild();
        updateSidebarMenuButtons();
        pagerController.showImmediate(currentPage);
        currentPage = pagerController.getCurrentPage();
    }

    public boolean onBackPressed() {
        if (sidebarController.onBackPressed()) {
            return true;
        }
        if (currentPage != PAGE_BOOKSHELF) {
            selectHomePage(PAGE_BOOKSHELF, true, false);
            return true;
        }
        return false;
    }

    public boolean handleTouchEvent(MotionEvent event) {
        if (sidebarController.handleTouchEvent(event)) {
            return true;
        }
        return effectiveMode == HomeNavigationResolver.MODE_BOTTOM
                && pagerController.handleTouchEvent(event);
    }

    public boolean consumePendingChildTouchCancel() {
        return sidebarController.consumePendingChildTouchCancel();
    }

    private void setup() {
        if (rootView != null) {
            rootView.addOnLayoutChangeListener((view, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
                if (right - left != oldRight - oldLeft || bottom - top != oldBottom - oldTop) {
                    refreshFromSettings();
                }
            });
        }
        refreshFromSettings();
    }

    private void bindSidebarMenuButtons() {
        bindSidebarMenuButton(sidebarBookshelfButton);
        bindSidebarMenuButton(sidebarStatsButton);
        bindSidebarMenuButton(sidebarBookmarksButton);
        bindSidebarMenuButton(sidebarSettingsButton);
    }

    private void bindSidebarMenuButton(View button) {
        if (button != null) {
            button.setOnClickListener(v -> sidebarController.openDrawer());
        }
    }

    private void updateSidebarMenuButtons() {
        int visibility = effectiveMode == HomeNavigationResolver.MODE_SLIDE_SIDEBAR ? View.VISIBLE : View.GONE;
        setVisibility(sidebarBookshelfButton, visibility);
        setVisibility(sidebarStatsButton, visibility);
        setVisibility(sidebarBookmarksButton, visibility);
        setVisibility(sidebarSettingsButton, visibility);
    }

    private void setVisibility(View view, int visibility) {
        if (view != null) {
            view.setVisibility(visibility);
        }
    }

    private void selectHomePage(int page, boolean animate, boolean syncFirst) {
        if (!activePages.contains(page)) {
            return;
        }
        if (sidebarController.isDrawerOpen()) {
            sidebarController.closeDrawer();
        }
        pagerController.selectPage(page, animate, syncFirst);
        currentPage = pagerController.getCurrentPage();
        updateNavigationSelection();
    }

    private void updateNavigationSelection() {
        bottomNavigationController.updateSelection(currentPage);
        sidebarController.updateSelection(currentPage);
    }

    private View sectionForPage(int page) {
        if (page == PAGE_STATS) return sectionReadingStats;
        if (page == PAGE_BOOKMARKS) return sectionBookmarks;
        if (page == PAGE_SETTINGS) return sectionSettings;
        if (page == PAGE_BOOKSHELF) return sectionBookshelf;
        return null;
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/HomeNavigationController.kt
package com.metahumanz.pacilread

import android.app.Activity
import android.view.MotionEvent
import android.view.View
import com.metahumanz.pacilread.storage.SettingsStore

class HomeNavigationController(
    private val activity: Activity,
    settingsStore: SettingsStore,
    private val callback: Callback,
) {
    interface Callback {
        fun isReadingTimeTrackingEnabled(): Boolean
        fun onHomePageSelected(page: Int, syncFirst: Boolean)
    }

    private val activePages = ArrayList<Int>()
    private val rootView: View? = activity.findViewById(R.id.main_root)
    private val pageContainer: View? = activity.findViewById(R.id.page_container)
    private val sectionBookshelf: View? = activity.findViewById(R.id.section_bookshelf)
    private val sectionReadingStats: View? = activity.findViewById(R.id.section_reading_stats)
    private val sectionBookmarks: View? = activity.findViewById(R.id.section_bookmarks)
    private val sectionSettings: View? = activity.findViewById(R.id.section_home_settings)
    private val loadingLayout: View? = activity.findViewById(R.id.layout_loading)
    private val sidebarBookshelfButton: View? = activity.findViewById(R.id.button_home_sidebar_bookshelf)
    private val sidebarStatsButton: View? = activity.findViewById(R.id.button_home_sidebar_stats)
    private val sidebarBookmarksButton: View? = activity.findViewById(R.id.button_home_sidebar_bookmarks)
    private val sidebarSettingsButton: View? = activity.findViewById(R.id.button_home_sidebar_settings)
    private val resolver = HomeNavigationResolver(activity, settingsStore, rootView)
    private val bottomNavigationController = HomeBottomNavigationController(
        activity,
        settingsStore,
        resolver,
        object : HomeBottomNavigationController.Callback {
            override fun onBottomPageSelected(page: Int) = selectHomePage(page, true, page == PAGE_STATS)
        },
    )
    private val sidebarController = HomeSidebarController(
        activity,
        settingsStore,
        resolver,
    ) { page, animate -> selectHomePage(page, animate, page == PAGE_STATS) }
    private val pagerController = HomePagerController(pageContainer, object : HomePagerController.Callback {
        override fun sectionForPage(page: Int): View? = this@HomeNavigationController.sectionForPage(page)

        override fun isTouchBlocked(event: MotionEvent): Boolean =
            loadingLayout == null || loadingLayout.visibility == View.VISIBLE ||
                bottomNavigationController.isTouchInBottomNavigation(event)

        override fun onPageChanged(page: Int, syncFirst: Boolean) {
            currentPage = page
            callback.onHomePageSelected(page, syncFirst || page == PAGE_STATS)
        }

        override fun onSelectionChanged() = updateNavigationSelection()
    })

    private var currentPage = PAGE_BOOKSHELF
    private var effectiveMode = HomeNavigationResolver.MODE_BOTTOM

    init {
        bindSidebarMenuButtons()
        setup()
    }

    fun getCurrentPage(): Int = currentPage

    fun restoreHomePage(page: Int) = selectHomePage(page, false, false)

    fun refreshFromSettings() {
        activePages.clear()
        activePages.addAll(resolver.activePages(callback.isReadingTimeTrackingEnabled()))
        val previousMode = effectiveMode
        effectiveMode = resolver.resolveEffectiveMode()
        if (!activePages.contains(currentPage)) currentPage = PAGE_BOOKSHELF
        pagerController.setActivePages(activePages)
        pagerController.setCurrentPage(currentPage)
        sidebarController.setActivePages(activePages)
        bottomNavigationController.update(activePages, currentPage, effectiveMode)
        sidebarController.updateShell(effectiveMode, previousMode != effectiveMode)
        sidebarController.rebuild()
        updateSidebarMenuButtons()
        pagerController.showImmediate(currentPage)
        currentPage = pagerController.currentPage
    }

    fun onBackPressed(): Boolean {
        if (sidebarController.onBackPressed()) return true
        if (currentPage != PAGE_BOOKSHELF) {
            selectHomePage(PAGE_BOOKSHELF, true, false)
            return true
        }
        return false
    }

    fun handleTouchEvent(event: MotionEvent): Boolean {
        if (sidebarController.handleTouchEvent(event)) return true
        return effectiveMode == HomeNavigationResolver.MODE_BOTTOM && pagerController.handleTouchEvent(event)
    }

    fun consumePendingChildTouchCancel(): Boolean = sidebarController.consumePendingChildTouchCancel()

    private fun setup() {
        rootView?.addOnLayoutChangeListener { _, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom ->
            if (right - left != oldRight - oldLeft || bottom - top != oldBottom - oldTop) refreshFromSettings()
        }
        refreshFromSettings()
    }

    private fun bindSidebarMenuButtons() {
        bindSidebarMenuButton(sidebarBookshelfButton)
        bindSidebarMenuButton(sidebarStatsButton)
        bindSidebarMenuButton(sidebarBookmarksButton)
        bindSidebarMenuButton(sidebarSettingsButton)
    }

    private fun bindSidebarMenuButton(button: View?) {
        button?.setOnClickListener { sidebarController.openDrawer() }
    }

    private fun updateSidebarMenuButtons() {
        val visibility = if (effectiveMode == HomeNavigationResolver.MODE_SLIDE_SIDEBAR) View.VISIBLE else View.GONE
        setVisibility(sidebarBookshelfButton, visibility)
        setVisibility(sidebarStatsButton, visibility)
        setVisibility(sidebarBookmarksButton, visibility)
        setVisibility(sidebarSettingsButton, visibility)
    }

    private fun setVisibility(view: View?, visibility: Int) {
        view?.visibility = visibility
    }

    private fun selectHomePage(page: Int, animate: Boolean, syncFirst: Boolean) {
        if (!activePages.contains(page)) return
        if (sidebarController.isDrawerOpen) sidebarController.closeDrawer()
        pagerController.selectPage(page, animate, syncFirst)
        currentPage = pagerController.currentPage
        updateNavigationSelection()
    }

    private fun updateNavigationSelection() {
        bottomNavigationController.updateSelection(currentPage)
        sidebarController.updateSelection(currentPage)
    }

    private fun sectionForPage(page: Int): View? = when (page) {
        PAGE_STATS -> sectionReadingStats
        PAGE_BOOKMARKS -> sectionBookmarks
        PAGE_SETTINGS -> sectionSettings
        PAGE_BOOKSHELF -> sectionBookshelf
        else -> null
    }

    companion object {
        const val PAGE_BOOKSHELF = 0
        const val PAGE_STATS = 1
        const val PAGE_BOOKMARKS = 2
        const val PAGE_SETTINGS = 3
    }
}
